refactor to es6 syntax with classes

hornta opened the pull request:
### Changes

Summary of changes:
- Refactor existing prototypes to ES6 classes. This includes getting rid of the `wrapPropertyMethod` which hides methods on these classes. Now it's possible for IDES to autodetect these methods.
- Added eslint to catch mistakes during refactoring and to maintain es6 syntaxes in the future.
- Improved formatting of JSDoc comments and removed some unnecessary tags like `@class`, `@function` and `@memberof` now when the classes are real classes and have real methods. All class methods which are forwarding all of their arguments (`...args`) still has their respective JSDoc `@param` tags which technically is wrong but since the underlying library works the way it does then it will have to be like that for now.
- Removed `moment` devDep
- Updated some testing dependencies like mocha and sinon.
- Added node 16 to the ci test/build matrix.
- Got rid of dependency on es6-promisify by using native Promise objects, much like es6-promisify did.
- Got rid of the >10s long test `should expire access token from cache by the expires_in setting` by using sinons fake timers to advance time.

### References
https://github.com/auth0/node-auth0/issues/303
https://github.com/auth0/node-auth0/pull/304

### Testing

Test logic has not been altered other than conforming to the linting rules and es6 class syntax.

It tooks me some time to figure out why the tests were not working initially so if any newcomers see this you will need to install OpenSSL 1 (NOT 3 because that is not yet supported by the `pem` package) and make sure you remove the CRLF EOL in users.json test data file.

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All existing and new tests complete without errors

adamjmcgrath commented:
👋 Hi @hornta - you've clearly done a lot of work here, so many thanks

My concern is whether we need a breaking change for dropping support for older, albeit EOL, versions of Node. If that's the case we may be reluctant to take this PR if we don't want to do a major release. But I'll check with the team and get back to you.

hornta commented:
> 👋 Hi @hornta - you've clearly done a lot of work here, so many thanks
> 
> My concern is whether we need a breaking change for dropping support for older, albeit EOL, versions of Node. If that's the case we may be reluctant to take this PR if we don't want to do a major release. But I'll check with the team and get back to you.

Yea I agree that it should maybe be a breaking change since there is not really a good way of determining the minimum supported node version at this moment. However there are a bunch of object spread syntaxes in the code base right now which came to be supported after classes were supported in node so introducing classes should not cause a major version bump but I think it's quite good to define the supported engines.

frederikprijck commented:
@hornta , our CI doesnt run on forks so no worries about it.

adamjmcgrath commented:
@hornta - thanks for raising this. Happy to review once you resolve the conflicts.

I've set up the CI to run on Node versions 8, 10, 12, 14 - as long as it passes on those (which es2017 should do) it should be fine.

> I think it's quite good to define the supported engines.

I agree, but I don't want to introduce a breaking change, so could you remove the engine requirements

hornta commented:
> Thanks for your work on this @hornta, I'll take a look to see if there are any other options and get back to you

Great. I aim to not let this rot away because when this is in then it will be easier converting to typescript because it uses classes then.

hornta commented:
> I'd rather avoid adding new options to this class just for testing.
> 
> If it comes to that we can always switch to something like jest, which has good options for stubbing esm

Alright so I reverted the changes so that proxyquire are used now. I now know why I wanted to remove proxyquire and it was because proxyquire throw an error when I tried to stub a `sinon.spy()`(because it's missing a `.prototype` property) and I don't know why it's like that, it must have been because of some updated dependency. Whatever I fixed it by attaching an empty prototype object `myStub.prototype = {}`. Not the prettiest but it doesn't throw now.

adamjmcgrath commented:
Thanks for your work on this @hornta 

I'm going to leave merging this until after the thanksgiving break - but I'll do it up shortly after

adamjmcgrath commented:
FYI the coverage changes are as a result of changing a bunch of `wrapPropertyMethod`'s to functions with body's

![image](https://user-images.githubusercontent.com/1299658/142601889-be7d1f78-9ec0-4e70-9838-019631aacd98.png)

adamjmcgrath commented:
Thanks @hornta - will do a release this week
